projects/igniteui-angular/src/lib/grids/state.directive.ts

```diff
@@ -0,0 +1,513 @@
+        paging: true,
+        cellSelection: true,
+        rowSelection: true
+    };
+
+    private state: IGridState;
+
+    /**
+     *  An object with options determining if a certain feature state should be saved.
+     *
+     * ```html
+     * <igx-grid [igxGridState]="options"></igx-grid>
+     * ```
+     * ```typescript
+     * public options = {selection: false, advancedFiltering: false};
+     * ```
+     */
+    @Input('igxGridState')
+    public get options(): IGridStateOptions {
+       return this._options;
+    }
+
+    public set options(value: IGridStateOptions) {
+        Object.assign(this._options, value);
+    }
+
+    /**
+     * @hidden
+     */
+    constructor(
+        @Host() @Optional() private grid: IgxGridComponent,
+        private resolver: ComponentFactoryResolver) { }
+
+    /**
+     * Gets the state of a feature or states of all grid features, unless a certain feature is disabled through the `options` property.
+     * @param `serialize` determines whether the returned object will be serialized to JSON string. Default value is false.
+     * @param `feature` string or array of strings determining the features which state to retrieve. If skipped, returns all.
+     * @returns Returns the serialized to JSON string IGridState object, or the non-serialized IGridState object.
+     * ```html
+     * <igx-grid [igxGridState]="options"></igx-grid>
+     * ```
+     * ```typescript
+     * @ViewChild(IgxGridStateDirective, { static: true }) public state;
+     * let state =  this.state.getState();
+     * ```
+     */
+    public getState(serialize = true, feature?: string | string[]): IGridState | string  {
+        let state: IGridState | string;
+        if (feature) {
+            state = {};
+            if (Array.isArray(feature)) {
+                feature.forEach(f => {
+                    state = Object.assign(state, this.getGridFeature(f));
+                });
+            } else {
+                state = this.getGridFeature(feature);
+            }
+        } else {
+            state = this.getAllGridFeatures();
+        }
+        if (serialize) {
+            state = JSON.stringify(state, this.stringifyCallback);
+            return state as string;
+        } else {
+            return state as IGridState;
+        }
+    }
+
+    /**
+     * Restores grid features' state based on the IGridState object passed as an argument.
+     * @param IGridState object to restore state from.
+     * @returns
+     * ```html
+     * <igx-grid [igxGridState]="options"></igx-grid>
+     * ```
+     * ```typescript
+     * @ViewChild(IgxGridStateDirective, { static: true }) public state;
+     * this.state.setState(gridState);
+     * ```
+     */
+    public setState(state: IGridState | string) {
+        if (typeof state === 'string') {
+            state = JSON.parse(state);
+        }
+        this.state = state as IGridState;
+        this.restoreGridState();
+        this.grid.cdr.detectChanges();
```

Comment:
Why is this change detection call needed?

Reply:
restorePaging and restoreRowSelection need it, otherwise Expression changed after chedked error is thrown. Since both need it, I moved it in the setState method.